collator/src/lib.rs

```diff
@@ -215,197 +210,204 @@ where
+				},
+				None => Vec::new(),
+			};
 
-		let proposer_future = factory.lock().init(&last_head.header);
-
-		let wait_to_announce = self.wait_to_announce.clone();
-
-		Box::pin(async move {
-			let proposer = proposer_future
-				.await
-				.map_err(|e| {
-					error!(
-						target: "cumulus-collator",
-						"Could not create proposer: {:?}",
-						e,
-					)
-				})
-				.ok()?;
-
-			let inherent_data = Self::inherent_data(
-				inherent_providers,
-				global_validation,
-				local_validation,
-				downward_messages,
-			)?;
-
-			let Proposal {
-				block,
-				storage_changes,
-				proof,
-			} = proposer
-				.propose(
-					inherent_data,
-					Default::default(),
-					//TODO: Fix this.
-					Duration::from_millis(500),
-					RecordProof::Yes,
-				)
-				.await
-				.map_err(|e| {
-					error!(
-						target: "cumulus-collator",
-						"Proposing failed: {:?}",
-						e,
-					)
-				})
-				.ok()?;
-
-			let proof = match proof {
-				Some(proof) => proof,
-				None => {
-					error!(
-						target: "cumulus-collator",
-						"Proposer did not return the requested proof.",
-					);
+			let new_validation_code = sp_io::storage::get(well_known_keys::NEW_VALIDATION_CODE);
 
+			Some(Collation {
+				//TODO: What should we do here?
+				fees: 0,
+				upward_messages,
+				new_validation_code: new_validation_code.map(Into::into),
+				head_data,
+				proof_of_validity: PoV { block_data },
+			})
+		})
+	}
+
+	async fn produce_candidate(
+		mut self,
+		_: PHash,
+		validation_data: ValidationData,
+	) -> Option<Collation> {
+		trace!(target: "cumulus-collator", "Producing candidate");
+
+		let last_head =
+			match Block::Header::decode(&mut &validation_data.persisted.parent_head.0[..]) {
+				Ok(x) => x,
+				Err(e) => {
+					error!(target: "cumulus-collator", "Could not decode the head data: {:?}", e);
 					return None;
 				}
 			};
 
-			let (header, extrinsics) = block.deconstruct();
+		if !self.check_block_status(last_head.hash()) {
+			return None;
+		}
 
-			// Create the parachain block data for the validators.
-			let b = ParachainBlockData::<Block>::new(header.clone(), extrinsics, proof);
+		let proposer_future = self.proposer_factory.lock().init(&last_head);
 
-			let mut block_import_params = BlockImportParams::new(BlockOrigin::Own, header);
-			block_import_params.body = Some(b.extrinsics().to_vec());
-			// Best block is determined by the relay chain.
-			block_import_params.fork_choice = Some(ForkChoiceStrategy::Custom(false));
-			block_import_params.storage_changes = Some(storage_changes);
+		let proposer = proposer_future
+			.await
+			.map_err(|e| {
+				error!(
+					target: "cumulus-collator",
+					"Could not create proposer: {:?}",
+					e,
+				)
+			})
+			.ok()?;
 
-			if let Err(err) = block_import
-				.lock()
-				.import_block(block_import_params, Default::default())
-			{
+		let inherent_data = self.inherent_data(
+			&validation_data,
+			// TODO get the downward messages
+			Vec::new(),
```

Comment:
what was the blocker here so far? Is it not passed in to the collation pipeline at the moment?

Reply:
Oh, should it? I assumed the collator would call the runtime API for fetching those

Reply:
Before it was passed to the collation. I think it would be nice to follow this pattern, as we pass `ValidationData` as well.

Reply:
Yeah, I also think it would be convenient. But I also think it would potentially be less flexible, because after all it's the collator who knows which messages it should pull, whereas the collation-generation doesn't necessary know about the specifics of the network.

Reply:
Wouldn't it need to fetch all downward messages anyway?

> whereas the collation-generation doesn't necessary know about the specifics of the network.

Can you explain what you mean?

Reply:
Potentially, it might be many, esp. if you include all the HRMP queues. But not only that, imagine a parachain that doesn't need HRMP, then it can just avoid requesting HRMP data at all. I think that it may be fair for XCMP as well. A conclusion I draw from this, is that the collator knows about the target chain (= "the network") better than the collation-generation can do (unless, we complicate the interface way more). That's why I say that the pull model is more flexible, because the collator has more information on when and how to query this data.

Reply:
Okay fine :)